File: projector/go/pkg/projector/projector.go
package projector

import (
	"encoding/json"
	"os"
	"path"
)

type Data struct {
	Projector map[string]map[string]string `json:"projector"`
}

type Projector struct {
	config Config
	data   Data
}

func CreateProjector(config *Config, data *Data) *Projector {
	return &Projector{
		config: *config,
		data:   *data,
	}
}

func (p *Projector) GetValueAll() map[string]string {
	current := p.config.Pwd
	previous := ""

	out := map[string]string{}

	paths := []string{}

	for current != previous {
		paths = append(paths, current)

		previous = current
		current = path.Dir(current)
	}

	for i := len(paths) - 1; i >= 0; i-- {
		if dir, ok := p.data.Projector[paths[i]]; ok {
			for key, value := range dir {
				out[key] = value
			}
		}
	}

	return out
}
func (p *Projector) GetValue(key string) (string, bool) {
	current := p.config.Pwd
	previous := ""

	out := ""
	found := false

	for current != previous {

		if dir, ok := p.data.Projector[current]; ok {
			if value, ok := dir[key]; ok {
				out = value
				found = true
				break
			}
		}

		previous = current
		current = path.Dir(current)

	}

	return out, found
}

func (p *Projector) SetValue(key, value string) {
	if _, ok := p.data.Projector[p.config.Pwd]; !ok {
		p.data.Projector[p.config.Pwd] = map[string]string{}
	}

	p.data.Projector[p.config.Pwd][key] = value
}

func (p *Projector) RemoveValue(key string) {
	if dir, ok := p.data.Projector[p.config.Pwd]; ok {
		delete(dir, key)
	}
}

func defaultProjector(config *Config) *Projector {
	return &Projector{
		config: *config,
		data:   Data{Projector: map[string]map[string]string{}},
	}
}

func NewProjector(config *Config) *Projector {
	if _, err := os.Stat(config.Config); err != nil {
		contents, err := os.ReadFile(config.Config)

		if err != nil {
			return defaultProjector(config)
		}

		var data Data

		err = json.Unmarshal(contents, &data)

		if err != nil {
			return defaultProjector(config)
		}

		return &Projector{
			config: *config,
			data:   data,
		}
	}

	return defaultProjector(config)
}

File: projector/deno/projector.ts
import type { Config } from "./config.ts";
import { existsSync } from "jsr:@std/fs";
import * as path from "jsr:@std/path";

export type Data = {
  projector: {
    // pwd
    [key: string]: {
      // key    value
      [key: string]: string;
    };
  };
};

const defaultData: Data = {
  projector: {},
};

export class Projector {
  constructor(private readonly config: Config, private readonly data: Data) {
  }
  static fromConfig(config: Config): Projector {
    if (!existsSync(config.configPath)) {
      return new Projector(config, defaultData);
    }

    let data: Data = defaultData;

    try {
      data = JSON.parse(Deno.readTextFileSync(config.configPath));
    } catch {
      data = defaultData;
    }

    return new Projector(config, data);
  }

  getValueAll(): Record<string, string> {
    let current = this.config.pwd;
    let previous = "";
    const paths = [];

    do {
      previous = current;
      paths.push(current);
      current = path.dirname(current);
    } while (current !== previous);

    return paths.reverse().reduce((acc, path) => {
      const value = this.data.projector[path];
      if (value) {
        Object.assign(acc, value);
      }
      return acc;
    }, {});
  }

  getValue(key: string): string | undefined {
    let current = this.config.pwd;
    let previous = "";
    let out: string | undefined = undefined;

    do {
      const value = this.data.projector[current]?.[key];
      if (value) {
        out = value;
        break;
      }
      previous = current;
      current = path.dirname(current);
    } while (current !== previous);

    return out;
  }

  setValue(key: string, value: string) {
    let directoryData = this.data.projector[this.config.pwd];
    if (!directoryData) {
      directoryData = {};
      this.data.projector[this.config.pwd] = directoryData;
    }

    directoryData[key] = value;
  }

  removeValue(key: string) {
    const directoryData = this.data.projector[this.config.pwd];
    if (directoryData) {
      delete directoryData[key];
    }
  }
}
